### PR TITLE
Kms

### PR DESCRIPTION
**Pull request by hagabor**
Added KMS Service with GenerateDataKey function only (with parameters only for KeyId and NumberOfBytes).

### CODE REVIEW

File: src/internal/kms.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+     * @param  {string} message - human readable error message
+     * @param  {string} code - A unique short code representing the error that was emitted
+     * @param  {string} operation - Name of the failed Operation
+     */
+    constructor(message: string, code: string, operation: string) {
+        super(message, code)
+        this.name = 'KMSServiceError'
+        this.operation = operation
+    }
+}
```
Comment (oleiade) [inline src/internal/kms.ts:143]:
```suggestion
}

/**
 *  KMSKeyLength describes possible key lenght values for KMS API data key operations.
 */
enum KMSKeyLength {
    KeySize256 = 32,
    KeySize512 = 64,
}
```

File: src/internal/kms.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            throw new KMSServiceError(
+                'An error occured on the server side',
+                'InternalServiceError',
+                operation
+            )
+        }
+    }
+}
+
+// TODO: create a Tags type
```
Comment (oleiade) [inline src/internal/kms.ts:103]:
```suggestion
```

File: src/internal/kms.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        this.Plaintext = Plaintext
+    }
+    static fromJSON(json: JSONObject) {
+        return new DataKeyResp(
+            json.CiphertextBlob as string,
+            json.KeyId as string,
+            json.Plaintext as string
+        )
+    }
+}
```
Comment (oleiade) [inline src/internal/kms.ts:126]:
```suggestion
/**
 * Class representing a data key
 */
export class DataKey {
    /**
     * The Amazon Resource Name (key ARN) of the KMS key that encrypted the data key.
     */
    id: string

    /**
     * The (base64-encoded) encrypted copy of the data key.
     */
    ciphertextBlob: string

    /**
     * The plaintext data key.
     * Use this data key to encrypt your data outside of KMS. Then, remove it from memory as soon as possible.
     */
    plaintext: string

    constructor(CiphertextBlob: string, KeyId: string, Plaintext: string) {
        this.ciphertextBlob = CiphertextBlob
        this.id = KeyId
        this.plaintext = Plaintext
    }

    static fromJSON(json: JSONObject) {
        return new DataKey(
            json.CiphertextBlob as string,
            json.KeyId as string,
            json.Plaintext as string
        )
    }
}
```

File: src/internal/kms.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * GenerateDataKey
+     * Used to generate data key with the KMS key defined
+     * @param {string} KeyId
+     * @param {number} NumberOfBytes
+     * @throws {KMSServiceError}
+     * @throws {InvalidSignatureError}
+     */
+    GenerateDataKey(KeyId: string, NumberOfBytes: number = 32): DataKeyResp | undefined {
+        const body = JSON.stringify({ KeyId: KeyId, NumberOfBytes : NumberOfBytes })
```
Comment (oleiade) [inline src/internal/kms.ts:46]:
```suggestion

    /**
     * GenerateDataKey returns a unique symmetric data key for use outside of AWS KMS.
     *
     * This operation returns a plaintext copy of the data key and a copy that is encrypted under a symmetric encryption KMS key that you specify.
     * The bytes in the plaintext key are random; they are not related to the caller or the KMS key.
     * You can use the plaintext key to encrypt your data outside of AWS KMS and store the encrypted data key with the encrypted data.
     *
     * To generate a data key, specify the symmetric encryption KMS key that will be used to encrypt the data key.
     * You cannot use an asymmetric KMS key to encrypt data keys.
     *
     * Used to generate data key with the KMS key defined
     * @param {string} id - Specifies the symmetric encryption KMS key that encrypts the data key. Use its key ID, key ARN, alias name, or alias ARN.
     * @param {KMKeySize} size - Specifies the length of the data key in bytes. For example, use the value 64 to generate a 512-bit data key (64 bytes is 512 bits). Default is 32, and generates a 256-bit data key.
     * @throws {KMSServiceError}
     * @throws {InvalidSignatureError}
     * @returns {DataKey} - The generated data key.
     */
    GenerateDataKey(id: string, size: KMSKeySize = KMSKeySize.Size256): DataKey | undefined {
        const body = JSON.stringify({ KeyId: id, NumberOfBytes: size })
```

File: src/internal/kms.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                ...this.commonHeaders,
+                'X-Amz-Target': `TrentService.GenerateDataKey`,
+            }
+        )
+        const res = http.request(this.method, signedRequest.url, body, {
+            headers: signedRequest.headers,
+        })
+        this._handle_error('GenerateDataKey', res)
+
+        return DataKeyResp.fromJSON(res.json() as JSONObject)
```
Comment (oleiade) [inline src/internal/kms.ts:63]:
```suggestion
        return DataKey.fromJSON(res.json() as JSONObject)
```

File: src/internal/kms.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
+import { JSONArray, JSONObject } from 'k6'
+import http, { RefinedResponse, ResponseType } from 'k6/http'
+
+import { AWSClient, AWSRequest } from './client'
+import { AWSError } from './error'
+import { AWSConfig } from './config'
+import { InvalidSignatureError, URIEncodingConfig } from './signature'
+import { v4 as uuidv4 } from 'uuid'
```
Comment (oleiade) [inline src/internal/kms.ts:8]:
```suggestion
```

File: src/internal/kms.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
+import { JSONArray, JSONObject } from 'k6'
```
Comment (oleiade) [inline src/internal/kms.ts:1]:
```suggestion
import { JSONObject } from 'k6'
```

**Review by oleiade — CHANGES_REQUESTED** [review]
Thank you so much for your contribution @hagabor 🎉 🙇🏻 Much appreciated!

This looks excellent indeed. I've left a bunch of suggestions that are mainly targeted towards improving type-safety, naming, documentation and hiding AWS internal naming conventions. Please let me know if you see potential further refinements, and incorporate them 🙇🏻

I will be testing this against our AWS setup and get back to you as soon as I have confirmed for myself that it works as expected.

Note that I'll be out of office for the upcoming week, and might be slower to repond in that time period 😉 

PS: Once we're good to go, I'll amend your PR with a regeneration of `.map` file as some type names will have changed 👍🏻   

Cheers!

**Comment by oleiade** [comment]
PR Checklist:
- [ ] @oleiade test against actual KMS

**Comment by oleiade** [comment]
I've tested the KMS feature you introduced @hagabor and everything seems to work as expected 🎉 Great job 🤝 

Please go ahead and integrate the suggested changes, and we should be good to merge this in preparation for a `v0.5.0` release. In order to stay productive, if I do not get an answer from you in the upcoming 7 days, I will go ahead and commit my suggestions and merge this so we don't stay blocked.

Cheers!

File: src/internal/kms.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            json.Plaintext as string
+        )
+    }
+}
+
+export class KMSServiceError extends AWSError {
+    operation: string
+
+    /**
+     * Constructs a SecretsManagerServiceError
```
Comment (hagabor) [inline src/internal/kms.ts:132]:
```suggestion
     * Constructs a KMSServiceError
```

**Comment by hagabor** [comment]
Dear Théo,

Sorry to responded with such delay. I'm going to integrate the suggested changes. 

brgds,
Gábor

File: src/kms.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+// Import only symbols we wish to re-export publicly
+import { signHeaders, InvalidSignatureError, URIEncodingConfig } from './internal/signature'
+import { AWSConfig, InvalidAWSConfigError } from './internal/config'
+import {
+    KMSClient,
+    DataKeyResp,
```
Comment (hagabor) [inline src/kms.ts:6]:
```suggestion
    DataKey,
```

File: src/kms.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    // AWS Signature V4
+    signHeaders,
+    InvalidSignatureError,
+    URIEncodingConfig,
+    // Aws Config
+    AWSConfig,
+    InvalidAWSConfigError,
+    // KMS
+    KMSClient,
+    DataKeyResp,
```
Comment (hagabor) [inline src/kms.ts:21]:
```suggestion
    DataKey,
```

**Comment by hagabor** [comment]
Dear Théo,

I've added some changes. One place I've left comment from the initial SecretsManager service (what I've used as starting point). Also changed export, and import symbols. Now seems working. Please review if I can ask you. Thanks!  

Brgds,
Gábor

**Comment by oleiade** [comment]
Hey @hagabor 

Thanks a lot for this, much appreciated 🙇🏻 
I'll merge this, and add end 2 end tests for it myself 👍🏻

**Pull request merged**